kernel/net/src/lib.rs

```diff
@@ -0,0 +1,61 @@
+#![no_std]
+
+extern crate alloc;
+
+use alloc::vec::Vec;
+use irq_safety::MutexIrqSafe;
+use smoltcp::wire::Ipv4Address;
+use spin::Mutex;
+
+mod device;
+mod error;
+mod interface;
+
+pub use device::{Device, DeviceCapabilities};
+pub use error::{Error, Result};
+pub use interface::{Interface, IpAddress, IpCidr, SocketSet};
+pub use smoltcp::{phy, socket, time::Instant, wire};
+
+/// A randomly chosen IP address that must be outside of the DHCP range.
+///
+/// The default QEMU user-slirp network gives IP address of `10.0.2.*`.
+const DEFAULT_LOCAL_IP: &str = "10.0.2.15/24";
+
+/// Standard home router address.
+///
+/// `10.0.2.2` is the default QEMU user-slirp networking gateway IP.
+const DEFAULT_GATEWAY_IP: IpAddress = IpAddress::Ipv4(Ipv4Address::new(10, 0, 2, 2));
```

Comment:
when we first implemented networking, smoltcp didn't support DHCP. I think it does now, so it'd be nice to replace these with those obtained from the DHCP reservation request. Could also save it for another PR but it's definitely something we'd need to do prior to supporting `std` in full.

Reply:
I quickly tried implementing it but it seems trickier than just polling a socket. I'd rather do it in a separate PR so I'll create an issue for it once this PR lands.